trin-core/src/portalnet/overlay.rs

```diff
@@ -152,20 +145,229 @@ impl Overlay {
     }
 
     /// Returns a vector of all ENR node IDs of nodes currently contained in the routing table.
-    pub fn table_entries_id(&self) -> Vec<NodeId> {
+    pub async fn table_entries_id(&self) -> Vec<NodeId> {
         self.kbuckets
             .write()
+            .await
             .iter()
             .map(|entry| *entry.node.key.preimage())
             .collect()
     }
 
     /// Returns a vector of all the ENRs of nodes currently contained in the routing table.
-    pub fn table_entries_enr(&self) -> Vec<Enr> {
+    pub async fn table_entries_enr(&self) -> Vec<Enr> {
         self.kbuckets
             .write()
+            .await
             .iter()
             .map(|entry| entry.node.value.enr().clone())
             .collect()
     }
+
+    pub async fn send_ping(&self, data_radius: U256, enr: Enr) -> Result<Vec<u8>, String> {
+        let enr_seq = self.discovery.read().await.local_enr().seq();
+        let msg = Ping {
+            enr_seq,
+            data_radius,
+        };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::Ping(msg)).to_bytes(),
+            )
+            .await
+    }
+
+    pub async fn send_find_nodes(&self, distances: Vec<u16>, enr: Enr) -> Result<Vec<u8>, String> {
+        let msg = FindNodes { distances };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::FindNodes(msg)).to_bytes(),
+            )
+            .await
+    }
+
+    pub async fn send_find_content(
+        &self,
+        content_key: Vec<u8>,
+        enr: Enr,
+    ) -> Result<Vec<u8>, String> {
+        let msg = FindContent { content_key };
+        self.discovery
+            .write()
+            .await
+            .send_talkreq(
+                enr,
+                PROTOCOL.to_string(),
+                Message::Request(Request::FindContent(msg)).to_bytes(),
+            )
+            .await
+    }
+}
+
+/// History network layer on top of the overlay protocol. Encapsulates history network specific data and logic.
+#[derive(Clone)]
+pub struct HistoryNetwork {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+/// State network layer on top of the overlay protocol. Encapsulates state network specific data and logic.
+#[derive(Clone)]
+pub struct StateNetwork {
+    pub overlay: Arc<OverlayProtocol>,
+}
+
+impl StateNetwork {
```

Comment:
I was still expecting these definitions to live in the respective crates, like `trin-state`. It feels backwards if we have to define every possible overlay definition in the core package, instead of only defining the shared logic in core, and the distinct logic in each crate. What are the challenges to keeping it split out?

Reply:
Yeah, good catch, I totally forgot to move those in their own crates :)